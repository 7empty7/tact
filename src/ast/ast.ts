import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {
    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRef = {
    kind: 'type_ref',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction)[]
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public' } | { type: 'get' };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attribute: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementCall = {
    kind: 'statement_call',
    id: number,
    expression: ASTOpCall | ASTOpCallStatic,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: string[],
    expression: ASTExpression,
    ref: ASTRef
}

//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementCall | ASTSTatementAssign;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementCall | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_call' || src.kind === 'statement_assign';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}