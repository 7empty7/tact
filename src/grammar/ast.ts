import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {

    static merge(...refs: ASTRef[]) {
        if (refs.length === 0) {
            throw Error('Cannot merge 0 refs');
        }
        let r = refs[0].#interval;
        for (let i = 1; i < refs.length; i++) {
            r = r.coverageWith(r, refs[i].#interval);
        }
        return new ASTRef(r);
    }

    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTString = {
    kind: 'string',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

export type ASTLvalueRef = {
    kind: 'lvalue_ref',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRefSimple = {
    kind: 'type_ref_simple',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

export type ASTTypeRefMap = {
    kind: 'type_ref_map',
    id: number,
    key: string,
    value: string,
    ref: ASTRef
}

export type ASTTypeRef = ASTTypeRefSimple | ASTTypeRefMap;

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction | ASTProgramImport)[]
}

export type ASTProgramImport = {
    kind: 'program_import',
    id: number,
    path: ASTString,
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string,
    message: boolean,
    prefix: number | null,
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    init: bigint | boolean | null | undefined,
    as: string | null,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction | ASTInitFunction | ASTReceive)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public', ref: ASTRef } | { type: 'get', ref: ASTRef } | { type: 'mutates', ref: ASTRef } | { type: 'extends', ref: ASTRef };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTReceive = {
    kind: 'def_receive',
    id: number,
    selector: { kind: 'simple', arg: ASTArgument } | { kind: 'fallback' } | { kind: 'comment', comment: ASTString } | { kind: 'bounce', arg: ASTArgument },
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

export type ASTInitFunction = {
    kind: 'def_init_function',
    id: number,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementExpression = {
    kind: 'statement_expression',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: ASTLvalueRef[],
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTCondition = {
    kind: 'statement_condition',
    id: number,
    expression: ASTExpression,
    trueStatements: ASTStatement[],
    falseStatements: ASTStatement[];
    elseif: ASTCondition | null,
    ref: ASTRef,
}

export type ASTStatementWhile = {
    kind: 'statement_while'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementUntil = {
    kind: 'statement_until'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementRepeat = {
    kind: 'statement_repeat'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}


//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementExpression | ASTSTatementAssign | ASTCondition | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew | ASTNull | ASTLvalueRef;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementExpression | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef | ASTNull | ASTCondition | ASTInitFunction | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat | ASTReceive | ASTLvalueRef | ASTString | ASTProgramImport;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_expression' || src.kind === 'statement_assign' || src.kind === 'statement_condition' || src.kind === 'statement_while' || src.kind === 'statement_until' || src.kind === 'statement_repeat';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function __DANGER_resetNodeId() {
    nextId = 1;
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}