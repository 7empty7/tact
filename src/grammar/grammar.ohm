Tact {

    // Starting point of the program
    Program = ProgramItem*
    ProgramItem = Struct
                | Contract
                | Primitive
                | StaticFunction
                | NativeFunction

    // Built-in declarations
    Primitive = primitive Type ";"

    // Static function
    StaticFunction = Function
    NativeFunction = nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ";" --withVoid
                   | nameAttribute "(" id ")" native id "(" ListOf<FunctionArg,","> ")" ":" Type ";" --withType
    
    // Field declarations
    Type = typeLiteral "?" --optional
         | typeLiteral --required
    Field = var id ":" Type ";"

    // Struct
    Struct = struct id "{" StructBody* "}"
    StructBody = Field

    // Contract
    Contract = contract id "{" ContractBody* "}"
    ContractBody = Field
                 | Function 

    // Function
    FunctionAttribute = public --public
                      | get    --getter
    Function = ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" "{" Statement* "}" --withVoid
             | ListOf<FunctionAttribute," "> fun id "(" ListOf<FunctionArg,","> ")" ":" Type "{" Statement* "}" --withType
             
    FunctionArg = id ":" Type

    // Statements
    Statement = StatementLet
              | StatementBlock
              | StatementReturn
              | StatementCall
              | StatementStaticCall
              | StatementAssign
    StatementBlock = "{" Statement* "}"
    StatementLet = let id ":" Type "=" Expression ";"
    StatementReturn = return Expression ";"
    StatementCall = ExpressionCall ";"
    StatementStaticCall = ExpressionStaticCall ";"
    StatementAssign = LValue Expression ";"

    // L-value
    LValue = id "=" --id
           | id "." LValue --subId

    // Expressions
    Expression = ExpressionOr
    ExpressionOr = ExpressionOr "||" ExpressionAnd --or
                 | ExpressionAnd
    ExpressionAnd = ExpressionAnd "&&" ExpressionCompare --and
                  | ExpressionCompare
    ExpressionCompare = ExpressionCompare "!=" ExpressionAdd --not
                      | ExpressionCompare "==" ExpressionAdd --eq
                      | ExpressionCompare ">" ExpressionAdd --gt
                      | ExpressionCompare ">=" ExpressionAdd --gte
                      | ExpressionCompare "<" ExpressionAdd --lt
                      | ExpressionCompare "<=" ExpressionAdd --lte
                      | ExpressionAdd
    ExpressionAdd = ExpressionAdd "+" ~"+" ExpressionMul --add
                  | ExpressionAdd "-" ~"-" ExpressionMul --sub
                  | ExpressionMul
    ExpressionMul = ExpressionMul "*" ExpressionUnary --mul
                  | ExpressionMul "/" ExpressionUnary --div
                  | ExpressionUnary
    ExpressionUnary = "-" ExpressionUnarySuffix --neg
                    | "+" ExpressionUnarySuffix --add
                    | "!" ExpressionUnarySuffix --not
                    | ExpressionUnarySuffix
    ExpressionUnarySuffix = ExpressionValue "!!" --notNull
                          | ExpressionValue
    ExpressionBracket = "(" Expression ")"

    // Order is important
    ExpressionValue = ExpressionCall
                    | ExpressionField
                    | ExpressionStaticCall
                    | ExpressionBracket
                    | ExpressionNew
                    | integerLiteral
                    | boolLiteral
                    | id
                    
    ExpressionField = ExpressionValue "." id
    ExpressionCall = ExpressionValue "." id "(" ListOf<Expression, ","> ")"
    ExpressionNew = id "{" ListOf<NewParameter, ","> "}"
    NewParameter = id ":" Expression
    ExpressionStaticCall = ExpressionValue "(" ListOf<Expression, ","> ")"

    // Type Literal
    typeLiteral = letterAsciiUC typeLiteralPart*
    typeLiteralPart = letterAscii | digit | "_"

    // Integer Literal
    // hexDigit defined in Ohm's built-in rules (otherwise: hexDigit = "0".."9" | "a".."f" | "A".."F")
    // digit defined in Ohm's built-in rules (otherwise: digit = "0".."9")
    integerLiteral = integerLiteralHex | integerLiteralDec // Order is important
    integerLiteralDec = digit+
    integerLiteralHex = "0x" hexDigit+
                      | "0X" hexDigit+

    // Letters
    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC

    // ID Literal
    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~reservedWord idStart idPart*

    // Bool Literal
    boolLiteral = ("true" | "false") ~idPart

    // Keywords
    keyword = struct | contract | var | fun | let | return | primitive | extend | native | public | get | null // Order is important
    struct = "struct" ~idPart
    contract = "contract" ~idPart
    var = "var" ~idPart
    let = "let" ~idPart
    fun = "fun" ~idPart
    return = "return" ~idPart
    primitive = "primitive" ~idPart
    extend = "extend" ~idPart
    native = "native" ~idPart
    public = "public" ~idPart
    null = "null" ~idPart
    get = "get" ~idPart

    // Attributes
    nameAttribute = "@name"

    // Reserved
    reservedWord = keyword
}
