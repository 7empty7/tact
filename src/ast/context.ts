import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}