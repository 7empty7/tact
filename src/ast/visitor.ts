import { ASTArgument, ASTBuiltIn, ASTContract, ASTExpression, ASTField, ASTFunction, ASTNode, ASTStatement, ASTStruct } from "./ast";
import { CompilerContext } from "./context";

export type ContextVisitor = {
    visit(ctx: CompilerContext, node: ASTNode): void;
    visitEnd(ctx: CompilerContext, node: ASTNode): void;
}

function visitDefField(ctx: CompilerContext, visitor: ContextVisitor, s: ASTField): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitArg(ctx: CompilerContext, visitor: ContextVisitor, s: ASTArgument): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx.addVariable({ name: s.name, node: s });
}

function visitExpression(ctx: CompilerContext, visitor: ContextVisitor, s: ASTExpression): CompilerContext {
    visitor.visit(ctx, s);
    if (s.kind === 'op_binary') {
        visitExpression(ctx, visitor, s.left);
        visitExpression(ctx, visitor, s.right);
    } else if (s.kind === 'op_unary') {
        visitExpression(ctx, visitor, s.right);
    } else if (s.kind === 'op_field') {
        visitExpression(ctx, visitor, s.src);
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitStatement(ctx: CompilerContext, visitor: ContextVisitor, s: ASTStatement): CompilerContext {
    visitor.visit(ctx, s);
    if (s.kind === 'let') {

        // Visit right hand side
        visitExpression(ctx, visitor, s.expression);

        // Add variable to context (AFTER visiting expression to avoid cyclic references)
        ctx = ctx.addVariable({ name: s.name, node: s });
    } else if (s.kind === 'return') {
        // Nothing to do
    } else {
        throw Error('Unknown type');
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitDefFunction(ctx: CompilerContext, visitor: ContextVisitor, s: ASTFunction): CompilerContext {
    visitor.visit(ctx, s);

    // Process all args
    for (let a of s.args) {
        ctx = visitArg(ctx, visitor, a);
    }

    // Process all statements
    for (let a of s.statements) {
        ctx = visitStatement(ctx, visitor, a);
    }

    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitStruct(ctx: CompilerContext, visitor: ContextVisitor, s: ASTStruct): CompilerContext {
    visitor.visit(ctx, s);
    for (let f of s.fields) {
        ctx = visitDefField(ctx, visitor, f);
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitContract(ctx: CompilerContext, visitor: ContextVisitor, s: ASTContract): CompilerContext {
    visitor.visit(ctx, s);

    // Add self to context
    ctx = ctx.addVariable({ name: 'self', node: s });

    // Process all declarations
    for (let d of s.declarations) {
        if (d.kind === 'def_field') {
            ctx = visitDefField(ctx, visitor, d);
        } else if (d.kind === 'def_function') {
            ctx = visitDefFunction(ctx, visitor, d);
        } else {
            throw Error('Unknown type');
        }
    }
    visitor.visitEnd(ctx, s);
    return ctx;
}

function visitBuiltIn(ctx: CompilerContext, visitor: ContextVisitor, s: ASTBuiltIn): CompilerContext {
    visitor.visit(ctx, s);
    visitor.visitEnd(ctx, s);
    return ctx;
}

export function visit(ctx: CompilerContext, visitor: ContextVisitor) {
    for (let n of Object.keys(ctx.astTypes)) {
        let t = ctx.astTypes[n];
        if (t.kind === 'def_struct') {
            ctx = visitStruct(ctx, visitor, t);
        } else if (t.kind === 'def_contract') {
            ctx = visitContract(ctx, visitor, t);
        } else if (t.kind === 'built-in') {
            ctx = visitBuiltIn(ctx, visitor, t);
        } else {
            throw Error('Unknown type');
        }
    }
}