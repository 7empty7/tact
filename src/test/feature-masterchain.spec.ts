import { toNano } from 'ton-core';
import { ContractSystem } from 'ton-emulator';
import { __DANGER_resetNodeId } from '../grammar/ast';
import { MasterchainTester } from './features/output/masterchain_MasterchainTester';
import { MasterchainTester as EnabledTester } from './features/output/masterchain-allow_MasterchainTester';

describe('feature-masterchain', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });

    //
    // Deployment and simple message receiving
    //

    it('should deploy to the workchain', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await MasterchainTester.fromInit());
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
    });
    it('should deploy to the workchain when masterchain enabled', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await EnabledTester.fromInit());
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
    });
    it('should not deploy to the workchain from masterchain', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure', -1);
        let contract = system.open(await MasterchainTester.fromInit());
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
    });
    it('should deploy to the workchain from masterchain when masterchain enabled', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure', -1);
        let contract = system.open(await EnabledTester.fromInit());
        let tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
    });

    //
    // newAddress
    //

    it('should create address for the workchain', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await MasterchainTester.fromInit());
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        await contract.getCreateAddress(0n, 0n);
    });

    it('should not create address for the materchain', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await MasterchainTester.fromInit());
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        await expect(contract.getCreateAddress(-1n, 0n)).rejects.toThrowError('Masterchain support is not enabled for this contract');
    });

    it('should create address for the masterchain when masterchain enabled', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await EnabledTester.fromInit());
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        await contract.getCreateAddress(-1n, 0n);
    });

    it('should not create address for invalid workchain', async () => {
        let system = await ContractSystem.create();
        let treasure = system.treasure('treasure');
        let contract = system.open(await MasterchainTester.fromInit());
        await contract.send(treasure, { value: toNano('10') }, "Deploy");
        await system.run();
        await expect(contract.getCreateAddress(10n, 0n)).rejects.toThrowError('Invalid address');
    });
});