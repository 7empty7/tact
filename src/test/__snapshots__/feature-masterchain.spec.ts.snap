// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature-masterchain should deploy to the workchain 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQAVoQ7tc9-XaPwcj5SQ8D0_KsRpwt86HQLAIfkA_Sye3OqL",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3890n,
      },
    ],
  },
]
`;

exports[`feature-masterchain should deploy to the workchain from masterchain when masterchain enabled 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBlRnC2dsBPOmABd7xF0qhfx2oYTm6Fxx7Iy9SNiPg5Haqh",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3674n,
      },
    ],
  },
]
`;

exports[`feature-masterchain should deploy to the workchain when masterchain enabled 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBlRnC2dsBPOmABd7xF0qhfx2oYTm6Fxx7Iy9SNiPg5Haqh",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3674n,
      },
    ],
  },
]
`;

exports[`feature-masterchain should not deploy to the workchain from masterchain 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQAVoQ7tc9-XaPwcj5SQ8D0_KsRpwt86HQLAIfkA_Sye3OqL",
          "type": "internal",
          "value": 10000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 137,
        "errorMessage": "Masterchain support is not enabled for this contract",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF000000004465706C6F79}",
            "type": "cell",
          },
          "bounce": false,
          "from": "kQAVoQ7tc9-XaPwcj5SQ8D0_KsRpwt86HQLAIfkA_Sye3OqL",
          "to": "@treasure(treasure)",
          "type": "internal",
          "value": 9988839000n,
        },
      },
    ],
  },
]
`;
